ttnn/cpp/ttnn/tensor/tensor_layout.cpp

```diff
@@ -0,0 +1,358 @@
+Size::Size(const std::array<size_t, 2>& size) : mHeight(size[0]), mWidth(size[1]) {}
+
+
+Size Size::operator/(const Size& rhs) const {
+    return Size(mHeight / rhs.mHeight, mWidth / rhs.mWidth);
+}
+
+Size Size::operator*(size_t scalar) const {
+    return Size(mHeight * scalar, mWidth * scalar);
+}
+
+Size::operator std::pair<size_t, size_t>() const {
+    return {mHeight, mWidth};
+}
+
+Size::operator std::array<size_t, 2>() const {
+    return {mHeight, mWidth};
+}
+
+Size::operator std::array<uint32_t, 2>() const {
+    return {static_cast<uint32_t>(mHeight), static_cast<uint32_t>(mWidth)};
+}
+
+size_t Size::height() const { return mHeight; }
+size_t Size::width() const { return mWidth; }
+
+bool Size::operator==(const Size& rhs) const {
+    return mHeight == rhs.mHeight && mWidth == rhs.mWidth;
+}
+
+// does not have to be a member, but it is easier to find if it is
+Size Size::aligned_to_tile(const Size& tile) {
+    size_t height = round_up(mHeight, tile.height());
+    size_t width = round_up(mWidth, tile.width());
+    return Size(height, width);
+}
+
+std::ostream& operator<<(std::ostream& os, const tt::tt_metal::Size& size)
+{
+    os << "(" << size.height() << ", " << size.width() << ")";
+    return os;
+}
+
+TensorLayout::TensorLayout(DataType dataType, const Size& tileSize, const MemoryConfig& memoryConfig, const Alignment& alignment)
+    : mDataType(dataType),
+      mTileSize(tileSize),
+      mMemoryConfig(memoryConfig),
+      mAlignment(alignment) {
+
+    mLayout = mTileSize.height() == 1 ? Layout::ROW_MAJOR : Layout::TILE;
+    initializeAlignment();
+    validateCustomAlignment();
+}
+
+TensorLayout::TensorLayout(DataType dataType, Layout layout, const MemoryConfig& memoryConfig, const Alignment& alignment)
+    : mLayout(layout),
+      mDataType(dataType),
+      mMemoryConfig(memoryConfig),
+      mAlignment(alignment) {
+
+    if (mLayout == Layout::TILE) {
+        mTileSize = Size(32, 32);
+    }
+    else {
+        mTileSize = Size(1, 1); // width 1 makes no sense? need juggestions
+    }
+
+    initializeAlignment();
+    validateCustomAlignment();
+}
+
+namespace {
+
+// This function is used to convert the legacy padded shape to Alignment
+// This is a temporary solution until we get rid of the LegacyPaddedShape
+// Keep in mind:
+//    The benefit of Alignment compared to LegacyPaddedShape is that it allows
+//    to pad between Channels and Batches with a row granularity.
+// Examples:
+// 1. Given LegacyPaddedShape (1, 1, 16, 20), alignment is (1, 1, 16, 20)
+//    means Shape (1, 1, 16, 16) which in theory takes physical space (1 * 1 * 16, 16)
+//    will take (1 * 1 * 16, 20)
+//
+// 2. Given LegacyPaddedShape (1, 1, 32, 32), alignment is (1, 1, 32, 32)
+//    Shape (1, 1, 16, 16)
+//      which in theory takes physical space (1 * 1 * 16, 16)
+//      will take (1 * 1 * 32, 32)
+//
+// 3. Given LegacyPaddedShape (2, 3, 32, 32), alignment is (2, 3*32, 32, 32).
```

Comment:
Shouldn't alignment for this case be `(2*3*32, 3*32, 32, 32)`?

Reply:
correct